orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/StageContext.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2017 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.pipeline.model;
+
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+import com.google.common.collect.ForwardingMap;
+
+public class StageContext extends ForwardingMap<String, Object> {
```

Comment:
This class is the key part of the change. It's a map implementation whose `get` method will look back using `Stage.ancestors()` for any missing keys.